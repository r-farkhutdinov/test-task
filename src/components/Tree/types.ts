export interface TreeProps {
    pages?: string[];
    isActive?: boolean;
    isActiveLevelTwo?: boolean;
}