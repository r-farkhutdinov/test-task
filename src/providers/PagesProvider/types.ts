import { ReactNode } from 'react';
import { TableData } from '../../types';

export type PagesContextValue = {
    data?: TableData;
    activePage?: string;
    query?: string;
    loading?: boolean;
    error?: string;
    changeActivePage?: (activePage: string) => void;
    setQuery?: (query: string) => void;
    load?: (q?: string) => void;
};

export type Props = {
    children: ReactNode;
}