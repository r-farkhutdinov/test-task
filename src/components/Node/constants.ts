export const basePadding = 22;
export const padding = 16;
export const nodeRootTestId = 'node-container';