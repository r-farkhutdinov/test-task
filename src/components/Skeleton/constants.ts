export const defaultLeftPadding = 16;
export const defaultRightPadding = 32;
export const skeletonRootTestId = 'skeleton';