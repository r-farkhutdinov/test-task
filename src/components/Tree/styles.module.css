.root {
    animation: animation 0.5s;
}

@keyframes animation {
    from { 
        opacity: 0; 
    }
    to { 
        opacity: 1; 
    }
}

.active { 
    background-color: var(--gray-first);
}

.activeLevelTwo { 
    background-color: var(--gray-second);
}
