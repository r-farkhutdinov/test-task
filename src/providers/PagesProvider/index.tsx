import { createContext, useCallback, useEffect, useMemo, useState } from 'react';
import { fetchData } from '../../api';
import type { TableData } from '../../types';
import type { PagesContextValue, Props } from './types';

export const PagesContext = createContext<PagesContextValue>({});

const errorMessage = 'Error fetching data. It is likely that the server is not running. Please start the server from its directory.';

/**
 * Context provider with the core logic of TOC: managing state, fetching data
 */
export const PagesProvider = ({ children }: Props) => {
    const [activePage, setActivePage] = useState<string>();
    const [data, setData] = useState<TableData>();
    const [query, setQuery] = useState('');
    const [error, setError] = useState<string>();
    const [loading, setLoading] = useState(false);

    const load = async (q?: string) => {
        setActivePage(undefined);
        try {
            setLoading(true);
            const result = await fetchData(q);
            setData(result);
        } catch {
            setError(errorMessage);
        } finally {
            setLoading(false);
        }
    };

    // this effect will fire twice in strict mode https://react.dev/reference/react/useEffect 
    useEffect(() => {
        load();
    }, []);

    const changeActivePage = useCallback((page: string) => {
        if (activePage !== page) {
            setActivePage(page);
        }
    }, [activePage]);

    const value = useMemo(() => ({
        activePage,
        data,
        query,
        loading,
        error,
        load,
        setQuery,
        changeActivePage
    }), [activePage, changeActivePage, data, error, loading, query]);

    return (
        <PagesContext.Provider value={value}>{children}</PagesContext.Provider>
    );
};