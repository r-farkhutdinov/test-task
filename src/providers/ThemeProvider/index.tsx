import { createContext, useCallback, useMemo, useState } from 'react';
import { storage } from '../../utils/storage';
import { type ThemeContextValue, type Props, Theme } from './types';

export const themeStorageKey = 'theme';

/**
 * Context provider with theme logic
 */
export const ThemeContext = createContext<ThemeContextValue>({
    theme: Theme.Light,
});

export const ThemeProvider = ({ children }: Props) => {
    const storedTheme = storage.get(themeStorageKey) as Theme;

    const [theme, setTheme] = useState(storedTheme ?? Theme.Light);

    const toggleTheme = useCallback(() => {
        const newTheme = theme === Theme.Light ? Theme.Dark : Theme.Light;
        setTheme(newTheme);
        storage.set(themeStorageKey, newTheme);
    }, [theme]);

    const value = useMemo(() => ({
        theme,
        toggleTheme
    }), [theme, toggleTheme]);

    return (
        <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
    );
};