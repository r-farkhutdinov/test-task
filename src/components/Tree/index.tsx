import { FC, useContext } from 'react';
import classNames from 'classnames/bind';
import { PagesContext } from '../../providers/PagesProvider';
import { getPageById } from '../../utils/getPageById';
import { Node } from '../Node';
import { rootTreeTestId } from './constants';
import type { TreeProps } from './types';
import styles from './styles.module.css';

const cx = classNames.bind(styles);

/**
 * Core tree components
 */
export const Tree: FC<TreeProps> = ({ pages, isActive, isActiveLevelTwo }) => {
    const { data } = useContext(PagesContext);

    const treeClassNames = cx(
        styles.root,
        {
            [styles.active]: isActive,
            [styles.activeLevelTwo]: isActiveLevelTwo,
        }
    );

    return (
        <div className={treeClassNames} data-testid={rootTreeTestId}>
            {pages?.map((page) => {
                const pageData = getPageById(data?.entities, page);
                return pageData ? <Node key={page} {...pageData} /> : null;
            })}
        </div>
    );
};
