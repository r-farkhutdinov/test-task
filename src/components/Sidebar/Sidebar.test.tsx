import { render, screen } from '@testing-library/react';
import { PagesContext } from '../../providers/PagesProvider';
import type { PagesContextValue } from '../../providers/PagesProvider/types';
import { rootTreeTestId } from '../Tree/constants';
import { Sidebar, noDataText } from '.';

const mockData = {
    topLevelIds: ['Page'],
    entities: {
        pages: {}
    }
};

const defaultContextValue: PagesContextValue = { data: mockData, loading: false };

const renderSidebarWithContext = (context: PagesContextValue) => render(
    <PagesContext.Provider value={context}>
        <Sidebar />
    </PagesContext.Provider>
);

describe('Sidebar component', () => {
    it('matches snapshot when loading', () => {
        const { asFragment } = renderSidebarWithContext({ ...defaultContextValue, loading: true });
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches snapshot when data is present', () => {
        const { asFragment } = renderSidebarWithContext(defaultContextValue);
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches snapshot with no data message when data is empty', () => {
        const { asFragment } = renderSidebarWithContext({ data: { ...mockData, topLevelIds: [] } });
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches snapshot when error', () => {
        const { asFragment } = renderSidebarWithContext({ ...defaultContextValue, error: 'Error' });
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render tree when data is present', () => {
        renderSidebarWithContext(defaultContextValue);
        expect(screen.getByTestId(rootTreeTestId)).toBeInTheDocument();
    });

    it('should render no data message when data is empty', () => {
        renderSidebarWithContext({ data: { ...mockData, topLevelIds: [] } });
        expect(screen.getByText(noDataText)).toBeInTheDocument();
    });

    it('should render error message on error', () => {
        renderSidebarWithContext({ ...defaultContextValue, error: 'Error' });
        expect(screen.getByText('Error')).toBeInTheDocument();
    });
});
