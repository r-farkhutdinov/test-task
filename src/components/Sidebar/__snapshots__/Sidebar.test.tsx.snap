// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component matches snapshot when data is present 1`] = `
<DocumentFragment>
  <aside
    class="root"
  >
    <div
      class="root"
      data-testid="tree"
    />
  </aside>
</DocumentFragment>
`;

exports[`Sidebar component matches snapshot when error 1`] = `
<DocumentFragment>
  <aside
    class="root"
  >
    <p
      class="error"
    >
      Error
    </p>
  </aside>
</DocumentFragment>
`;

exports[`Sidebar component matches snapshot when loading 1`] = `
<DocumentFragment>
  <aside
    class="root"
  >
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 0px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 16px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 32px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 0px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 16px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 32px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 0px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 16px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 32px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 0px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 16px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 32px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 0px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 16px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 32px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 0px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 16px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 32px; margin-right: 32px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 0px; margin-right: 0px;"
    />
    <div
      class="root"
      data-testid="skeleton"
      style="margin-left: 16px; margin-right: 32px;"
    />
  </aside>
</DocumentFragment>
`;

exports[`Sidebar component matches snapshot with no data message when data is empty 1`] = `
<DocumentFragment>
  <aside
    class="root"
  >
    <p
      class="error"
    >
      No data, please reset the filter
    </p>
  </aside>
</DocumentFragment>
`;
